### PR TITLE
best N hits

### PR DESCRIPTION
**Pull request by tyleraland**
New argument `--best N` keeps only the best N hits for each query sequence, as ranked by absolute number of mismatches from the query.

I've been comparing the before and after (with and without `--best 2`) for a particular data set generated by the capture pipeline (classifying individual query sequences) and the results appear pretty promising.  You can compare the before and after classifications here:

/mnt/disk2/molmicro/working/tland9/2016-02-25_bioy_slashname_filtering/classifications-nobest
/mnt/disk2/molmicro/working/tland9/2016-02-25_bioy_slashname_filtering/classifications-best2

Potential Feedback:

* Suggestions for a more descriptive argument name?
* Note the requirement on the "mismatches" field means blast results need to look a little different
* Some of my lines may be a little long.  Is there a strict 80 character style limit?

### CODE REVIEW

**Comment by nhoffman** [comment]
Can we name the argument --best-n-hits ?

File: bioy_pkg/subcommands/classifier.py [hunk 1]
```diff
@@ -561,6 +561,12 @@ def build_parser(parser):
         (NOT IMPLEMENTED)""")
     parser.add_argument(
         '--limit', type=int, help='limit number of blast results')
+    parser.add_argument(
+        '--best-n-hits', type=int, default=0,
```
Comment (crosenth) [inline bioy_pkg/subcommands/classifier.py]: Don't include a default here.  Later, check if argument has been set.

File: bioy_pkg/subcommands/classifier.py [hunk 2]
```diff
@@ -572,6 +578,8 @@ def action(args):
     names = None if args.has_header else sequtils.BLAST_HEADER_DEFAULT
     header = 0 if args.has_header else None
     usecols = ['qseqid', 'sseqid', 'pident', 'qcovs']
+    if args.best_n_hits > 0:
```
Comment (crosenth) [inline bioy_pkg/subcommands/classifier.py]: if args.best_n_hits:

File: bioy_pkg/subcommands/classifier.py [hunk 3]
```diff
@@ -707,11 +715,29 @@ def action(args):
         assignment_columns += blast_results_columns.tolist()
         blast_results = pd.DataFrame(columns=assignment_columns)
     else:
+
         blast_results_post_len = len(blast_results)
         log.info('{} ({:.0%}) valid hits selected'.format(
             blast_results_post_len,
             blast_results_post_len / blast_results_len))
 
+        if args.best_n_hits > 0:
```
Comment (crosenth) [inline bioy_pkg/subcommands/classifier.py]: if args.best_n_hits:

File: bioy_pkg/subcommands/classifier.py [hunk 4]
```diff
@@ -707,11 +715,29 @@ def action(args):
...
+            blast_results_len = len(blast_results)
+            query_groups = blast_results.groupby(by=['specimen', 'qseqid'])
+
+            # For each query, determine the mismatches of the Nth best hit
+            mismatch_threshold = query_groups.apply(lambda g: g['mismatch'].nsmallest(args.best_n_hits).iloc[-1])
```
Comment (crosenth) [inline bioy_pkg/subcommands/classifier.py]: I would just define a function that returns what you want in the groupby.apply:

def filter_mismatches(df, thresholh):
  if df['mismatch'].nsmallest(args.best_n_hits).iloc[-1]) >= threshold:
     return df

No you don't need the extra apply at line 732.

**Comment by crosenth** [comment]
Don't forget to update the CHANGLOG.rst

**Pull request merged**